File: civitasruby/lib/vista_textual.rb
#encoding:utf-8
require 'io/console'
require_relative 'civitas_juego'
require_relative 'operacion_inmobiliaria'
require_relative 'operaciones_juego'
require_relative 'respuestas'
require_relative 'gestiones_inmobiliarias'
require_relative 'salidas_carcel'

module Civitas

  class Vista_textual
    
    attr_reader:iGestion,:iPropiedad
    def mostrar_estado(estado)
      puts estado
    end

    
    def pausa
      print "Pulsa una tecla"
      STDIN.getch
      print "\n"
    end

    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)
            numero = cadena.to_i
            ok = true
          else
            raise IOError
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end



    def menu(titulo,lista)
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts tab+index.to_s+"-"+l
        index += 1
      }

      opcion = lee_entero(lista.length,
                          "\n"+tab+"Elige una opcion: ",
                          tab+"Valor erróneo")
      return opcion
    end

    def salir_carcel
      lista_salidas=[Civitas::SalidasCarcel::PAGANDO.to_s,
        Civitas::SalidasCarcel::TIRANDO.to_s]
      opcion=menu("Elige la forma para intentar salir de la carcel",
        lista_salidas)
      return Civitas::SalidasCarcel::Lista_salidas[opcion]
    end
    
    
    def comprar
      lista_respuestas=[Civitas::Respuestas::SI.to_s,
        Civitas::Respuestas::NO.to_s]
      opcion=menu("Desea usted comprar la calle a la que se ha llegado",
        lista_respuestas)
      return Civitas::Respuestas::Lista_respuestas[opcion]
    end

    def gestionar
      lista_gestiones=[Civitas::GestionesInmobiliarias::VENDER.to_s,
        Civitas::GestionesInmobiliarias::HIPOTECAR.to_s,
        Civitas::GestionesInmobiliarias::CANCELAR_HIPOTECA.to_s,
        Civitas::GestionesInmobiliarias::CONSTRUIR_CASA.to_s,
        Civitas::GestionesInmobiliarias::CONSTRUIR_HOTEL.to_s,
        Civitas::GestionesInmobiliarias::TERMINAR.to_s]
      opcion=menu("¿Que deseas hacer?",lista_gestiones)
      @iGestion=opcion
      n=@juegoModel.get_jugador_actual.get_nombre_propiedades
      if(opcion!=5 && n!=nil)
        indice=menu("¿Sobre que propiedad quieres realizar la gestion 
        elegida?",n)
      end
      @iPropiedad=indice
    end


    def mostrar_siguiente_operacion(operacion)
      puts operacion
    end

    def mostrarEventos
      while Diario.instance.eventos_pendientes
        puts Diario.instance.leer_evento+"\n"
      end
      
    end

    def set_civitas_juego(civitas)
         @juegoModel=civitas
         
    end

    def actualizar_vista
      a=@juegoModel.info_jugador_texto
      b=@juegoModel.get_casilla_actual.to_string
      puts a+"\n"+b+"\n"
    end

    
  end

end

File: JavaCivitas/src/juegoTexto/VistaTextual.java
package juegoTexto;

import civitas.CivitasJuego;
import civitas.Diario;
import civitas.OperacionesJuego;
import civitas.SalidasCarcel;
import civitas.Respuestas;
import civitas.GestionesInmobiliarias;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import civitas.Casilla;
import civitas.Jugador;
import civitas.TituloPropiedad;

public class VistaTextual {
  
  CivitasJuego juegoModel; 
  int iGestion=-1;
  int iPropiedad=-1;
  private static String separador = "=====================";
  
  private Scanner in;
  
  VistaTextual () {
    in = new Scanner (System.in);
  }
  
  void mostrarEstado(String estado) {
    System.out.println (estado);
  }
              
  void pausa() {
    System.out.print ("\nPulsa una tecla\n");
    in.nextLine();
  }

  int leeEntero (int max, String msg1, String msg2) {
    Boolean ok;
    String cadena;
    int numero = -1;
    do {
      System.out.print (msg1);
      cadena = in.nextLine();
      try {  
        numero = Integer.parseInt(cadena);
        ok = true;
      } catch (NumberFormatException e) { // No se ha introducido un entero
        System.out.println (msg2);
        ok = false;  
      }
      if (ok && (numero < 0 || numero >= max)) {
        System.out.println (msg2);
        ok = false;
      }
    } while (!ok);

    return numero;
  }

  int menu (String titulo, ArrayList<String> lista) {
    String tab = "  ";
    int opcion;
    System.out.println (titulo);
    for (int i = 0; i < lista.size(); i++) {
      System.out.println (tab+i+"-"+lista.get(i));
    }

    opcion = leeEntero(lista.size(),
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo");
    return opcion;
  }

  SalidasCarcel salirCarcel() {
    int opcion = menu ("\nElige la forma para intentar salir de la carcel", new ArrayList<> (Arrays.asList("Pagando","Tirando el dado")));
    return (SalidasCarcel.values()[opcion]);
  }

  Respuestas comprar() {

    ArrayList<String> lista = new ArrayList<>();
    lista.add("SI");
    lista.add("NO");

    int eleccion=menu("\n¿Desea comprar la calle a la que ha llegado?",lista);
    if(eleccion == 0){

      return Respuestas.SI;

    }else{

      return Respuestas.NO;

    }

  }

  void gestionar () {
      Jugador jugador = juegoModel.getJugadorActual();
      int opcion = menu ("¿Que deseas gestionar?",
       new ArrayList<> (Arrays.asList("VENDER", "HIPOTECAR", "CANCELAR_HIPOTECA", "CONSTRUIR_CASA", "CONSTRUIR_HOTEL", "TERMINAR")));
      iGestion = opcion;
      iPropiedad=jugador.getNumCasillaActual();
      //funcion getNumCasillaActual de jugador cambiada a publica.
  }
  
  public int getGestion(){
      return iGestion;
  }
  
  public int getPropiedad(){
      return iPropiedad;
  }
    

  void mostrarSiguienteOperacion(OperacionesJuego operacion) {
      System.out.print(operacion);
  }


  void mostrarEventos() {
      //cambiado el Diario leer eventos pendientes
      if(Diario.getInstance().eventosPendientes()){
        String cad = Diario.getInstance().leerEvento();
        System.out.print(cad);
      }
  }
  
  public void setCivitasJuego(CivitasJuego civitas){ 
        juegoModel=civitas;
        //this.actualizarVista();

    }
  
  void actualizarVista(){

      juegoModel.getJugadorActual().toString();
      System.out.println(juegoModel.getCasillaActual());
  } 
}
